File: orange-pi-daemon/src/sms_sender.zig
const std = @import("std");
const types = @import("types.zig");
const ModemManager = @import("modem_manager.zig").ModemManager;
const ApiClient = @import("api_client.zig").ApiClient;

/// Represents an outgoing SMS message
pub const OutgoingSMS = struct {
    id: []const u8,
    recipient: []const u8,         // Number to send to
    phone_iccid: []const u8,       // ICCID of the phone to send from
    content: []const u8,
    created_at: []const u8,
    
    pub fn deinit(self: *OutgoingSMS, allocator: std.mem.Allocator) void {
        allocator.free(self.id);
        allocator.free(self.recipient);
        allocator.free(self.phone_iccid);
        allocator.free(self.content);
        allocator.free(self.created_at);
    }
};

/// SMS sender responsible for fetching and sending SMS messages
pub const SMSSender = struct {
    allocator: std.mem.Allocator,
    api_client: *ApiClient,
    modem_manager: *ModemManager,
    config: types.Config,
    
    pub fn init(allocator: std.mem.Allocator, api_client: *ApiClient, modem_manager: *ModemManager, config: types.Config) SMSSender {
        return .{
            .allocator = allocator,
            .api_client = api_client,
            .modem_manager = modem_manager,
            .config = config,
        };
    }
    
    /// Get pending SMS messages from API
    pub fn getPendingSMS(self: *SMSSender) ![]OutgoingSMS {
        // Use the API client to get pending SMS
        const pending_sms = self.api_client.getPendingSms() catch |err| {
            // Only log in non-test builds
            if (@import("builtin").is_test == false) {
                std.log.err("Failed to get pending SMS: {any}", .{err});
            }
            return &[_]OutgoingSMS{};
        };
        defer {
            // Free the pending SMS data
            for (pending_sms) |sms| {
                self.allocator.free(sms.id);
                self.allocator.free(sms.phone_iccid);
                if (sms.phone_number) |pn| self.allocator.free(pn);
                self.allocator.free(sms.content);
                self.allocator.free(sms.recipient);
                self.allocator.free(sms.created_at);
            }
            self.allocator.free(pending_sms);
        }
        
        // Convert PendingSms to OutgoingSMS
        var sms_list = std.ArrayList(OutgoingSMS).init(self.allocator);
        defer sms_list.deinit();
        
        for (pending_sms) |sms| {
            const outgoing = OutgoingSMS{
                .id = try self.allocator.dupe(u8, sms.id),
                .recipient = try self.allocator.dupe(u8, sms.recipient),
                .phone_iccid = try self.allocator.dupe(u8, sms.phone_iccid),
                .content = try self.allocator.dupe(u8, sms.content),
                .created_at = try self.allocator.dupe(u8, sms.created_at),
            };
            try sms_list.append(outgoing);
        }
        
        // Convert to owned slice
        const result = try sms_list.toOwnedSlice();
        std.log.info("📱 Found {d} pending SMS messages", .{result.len});
        return result;
    }
    
    /// Find modem ID for a given ICCID
    pub fn findModemForIccid(self: *SMSSender, target_iccid: []const u8) ?[]const u8 {
        // Get list of modems
        const modems = self.modem_manager.listModems() catch return null;
        defer {
            for (modems) |modem| self.allocator.free(modem);
            self.allocator.free(modems);
        }
        
        // Check each modem for matching ICCID
        for (modems) |modem_id| {
            const iccid_opt = self.modem_manager.getIccid(modem_id) catch continue;
            if (iccid_opt) |iccid| {
                defer self.allocator.free(iccid);
                if (std.mem.eql(u8, iccid, target_iccid)) {
                    return self.allocator.dupe(u8, modem_id) catch null;
                }
            }
        }
        
        return null;
    }
    
    /// Send an SMS message
    pub fn sendSMS(self: *SMSSender, sms: OutgoingSMS) !void {
        // Find the modem for this ICCID
        const modem_id = self.findModemForIccid(sms.phone_iccid) orelse {
            // Only log in non-test builds
            if (@import("builtin").is_test == false) {
                std.log.err("No modem found for ICCID: {s}", .{sms.phone_iccid});
            }
            try self.reportSMSResult(sms.id, false, "No modem found for ICCID");
            return error.ModemNotFound;
        };
        defer self.allocator.free(modem_id);
        
        std.log.info("📤 Sending SMS from modem {s} to {s}", .{ modem_id, sms.recipient });
        
        // Send the SMS using ModemManager
        const sms_id = self.modem_manager.sendSms(modem_id, sms.recipient, sms.content) catch |err| {
            std.log.err("Failed to send SMS: {any}", .{err});
            try self.reportSMSResult(sms.id, false, "Failed to send SMS");
            return err;
        };
        
        // Report success
        try self.reportSMSResult(sms.id, true, "SMS sent successfully");
        std.log.info("✅ SMS sent successfully to {s} (SMS ID: {s})", .{ sms.recipient, sms_id });
    }
    
    /// Report SMS result back to API
    pub fn reportSMSResult(self: *SMSSender, message_id: []const u8, success: bool, message: []const u8) !void {
        _ = message;
        
        // Only mark as sent if actually successful
        if (success) {
            // Mark the SMS as sent using the API client
            self.api_client.markSmsAsSent(message_id) catch |err| {
                std.log.err("Failed to mark SMS as sent: {any}", .{err});
                return err;
            };
        } else {
            // TODO: Add API method to mark SMS as failed
            if (@import("builtin").is_test == false) {
                std.log.warn("SMS {s} failed to send - not marking as sent", .{message_id});
            }
        }
    }
};

File: orange-pi-daemon/src/sms_sender.rs
use anyhow::{anyhow, Result};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::Arc;
use tracing::{debug, error, info, warn};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PendingSms {
    pub id: String,
    pub recipient: String,
    pub phone_iccid: String,
    pub content: String,
    #[serde(default = "default_sms_purpose")]
    pub purpose: String,
    pub created_at: String,
}

fn default_sms_purpose() -> String {
    "user".to_string()
}

#[derive(Debug, Serialize)]
struct SmsResult {
    message_id: String,
    success: bool,
    error_message: Option<String>,
}

pub struct SmsSender {
    api_client: crate::api_client::ApiClient,
    modem_manager: Arc<crate::modem_manager::ModemManager>,
    modem_cache: HashMap<String, String>, // ICCID -> modem_id mapping
}

impl SmsSender {
    pub fn new(
        api_client: crate::api_client::ApiClient,
        modem_manager: Arc<crate::modem_manager::ModemManager>,
    ) -> Self {
        Self {
            api_client,
            modem_manager,
            modem_cache: HashMap::new(),
        }
    }

    /// Update the modem cache with current ICCID mappings
    pub fn update_modem_cache(&mut self, cache: HashMap<String, String>) {
        self.modem_cache = cache;
    }

    /// Fetch pending SMS messages from the API
    pub async fn get_pending_sms(&self) -> Result<Vec<PendingSms>> {
        let url = format!("{}/api/control/pending-sms", self.api_client.config.api_url);

        debug!("🌐 Fetching pending SMS from: {}", url);

        let client = reqwest::Client::new();
        let response = client
            .get(&url)
            .header("X-API-Key", &self.api_client.config.api_key)
            .header("Accept", "application/json")
            .timeout(std::time::Duration::from_secs(10))
            .send()
            .await?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            return Err(anyhow!(
                "Failed to fetch pending SMS: {} - {}",
                status,
                text
            ));
        }

        #[derive(Deserialize)]
        struct ApiResponse {
            pending_messages: Vec<PendingSms>,
        }

        let api_response: ApiResponse = response.json().await?;

        if !api_response.pending_messages.is_empty() {
            info!(
                "📱 Found {} pending SMS messages to send",
                api_response.pending_messages.len()
            );
        }

        Ok(api_response.pending_messages)
    }

    /// Find modem ID for a given ICCID using cache or D-Bus
    pub async fn find_modem_for_iccid(&self, target_iccid: &str) -> Option<String> {
        debug!("🔍 Searching for modem with ICCID: {}", target_iccid);

        // First check our cache
        if let Some(modem_id) = self.modem_cache.get(target_iccid) {
            debug!(
                "✅ Found modem {} in cache for ICCID {}",
                modem_id, target_iccid
            );
            return Some(modem_id.clone());
        }

        // If not in cache, search through all modems via D-Bus
        info!(
            "📱 ICCID {} not in cache, searching via D-Bus...",
            target_iccid
        );

        // List all modems using D-Bus
        let modem_ids = match self.modem_manager.list_modems().await {
            Ok(ids) => ids,
            Err(e) => {
                error!("❌ Failed to list modems via D-Bus: {}", e);
                return None;
            }
        };

        debug!("📱 Found {} modems to check", modem_ids.len());

        // Check each modem for matching ICCID
        for modem_id in modem_ids {
            match self.modem_manager.get_iccid(&modem_id).await {
                Ok(Some(iccid)) => {
                    debug!(
                        "📱 Modem {} has ICCID: {} (target: {})",
                        modem_id, iccid, target_iccid
                    );

                    if iccid == target_iccid {
                        info!("✅ Found modem {} for ICCID {}", modem_id, target_iccid);
                        return Some(modem_id);
                    }
                }
                Ok(None) => {
                    debug!("📱 Modem {} has no SIM", modem_id);
                }
                Err(e) => {
                    debug!("⚠️  Failed to get ICCID for modem {}: {}", modem_id, e);
                }
            }
        }

        error!("❌ No modem found with ICCID {}", target_iccid);
        None
    }

    /// Send an SMS message using ModemManager (AT commands or D-Bus)
    pub async fn send_sms(&self, sms: &PendingSms) -> Result<()> {
        // Find the modem for this ICCID
        let modem_id = match self.find_modem_for_iccid(&sms.phone_iccid).await {
            Some(id) => id,
            None => {
                error!("No modem found for ICCID: {}", sms.phone_iccid);
                let _ = self
                    .report_sms_result(&sms.id, false, Some("No modem found for ICCID"))
                    .await;
                return Err(anyhow!("ModemNotFound"));
            }
        };

        info!(
            "📤 Sending SMS from modem {} to {}",
            modem_id, sms.recipient
        );

        // Send SMS using ModemManager (supports both AT commands and D-Bus)
        match self
            .modem_manager
            .send_sms_with_short_code(
                &modem_id,
                &sms.recipient,
                &sms.content,
                sms.purpose == "balance_maintenance",
            )
            .await
        {
            Ok(_) => {
                // Report success
                self.report_sms_result(&sms.id, true, None).await?;
                info!(
                    "✅ SMS sent successfully to {} (Message ID: {})",
                    sms.recipient, sms.id
                );
                Ok(())
            }
            Err(e) => {
                let error_msg = format!("Failed to send SMS: {}", e);
                error!("{}", error_msg);
                self.report_sms_result(&sms.id, false, Some(&error_msg))
                    .await?;
                Err(anyhow!(error_msg))
            }
        }
    }

    /// Report SMS result back to API
    pub async fn report_sms_result(
        &self,
        message_id: &str,
        success: bool,
        error_message: Option<&str>,
    ) -> Result<()> {
        let url = format!("{}/api/control/sms-result", self.api_client.config.api_url);

        let result = SmsResult {
            message_id: message_id.to_string(),
            success,
            error_message: error_message.map(|s| s.to_string()),
        };

        debug!(
            "📤 Reporting SMS result for {}: success={}",
            message_id, success
        );

        let client = reqwest::Client::new();
        let response = client
            .post(&url)
            .header("Content-Type", "application/json")
            .header("X-API-Key", &self.api_client.config.api_key)
            .json(&result)
            .timeout(std::time::Duration::from_secs(10))
            .send()
            .await?;

        if !response.status().is_success() {
            let status = response.status();
            let text = response.text().await.unwrap_or_default();
            warn!("Failed to report SMS result: {} - {}", status, text);
            return Err(anyhow!(
                "Failed to report SMS result: {} - {}",
                status,
                text
            ));
        }

        if success {
            debug!("✅ Marked SMS {} as sent", message_id);
        } else {
            debug!("⚠️  Marked SMS {} as failed", message_id);
        }

        Ok(())
    }

    /// Process all pending SMS messages
    pub async fn process_pending_sms(&self) -> Result<()> {
        let pending_messages = self.get_pending_sms().await?;

        if pending_messages.is_empty() {
            return Ok(());
        }

        info!(
            "📤 Processing {} pending SMS messages",
            pending_messages.len()
        );

        for sms in pending_messages {
            match self.send_sms(&sms).await {
                Ok(_) => {
                    info!("✅ Successfully sent SMS to {}", sms.recipient);
                }
                Err(e) => {
                    error!("❌ Failed to send SMS to {}: {}", sms.recipient, e);
                    // Continue with next message even if this one failed
                }
            }
        }

        Ok(())
    }
}
